OCP-OCP migration (#411)
kubev2v/forklift-documentation: documentation/modules/creating-storage-mapping.adoc

You can create a storage mapping by using the {ocp} web console to map source disk storages to {virt} storage classes.

[NOTE]
====

> You can specify the source and target providers but leave the specific storage mappings empty for future use if VMs currently have no storages. This kind of mapping is called an _empty mapping_.

Reviewer: same here - there's nothing special about these empty mappings, they are just a concrete case
in https://github.com/kubev2v/forklift-console-plugin/pull/498 we enabled to specify empty mappings since VMs in RHV that only have LUN disks, have no storage domains to map to storage classes - I'd put a note in the UI and CLI sections that talk about plan creation, that say that storage/network mappings should always be specified when creating a plan even when the migrated VMs have no network/storage

Reviewer: actually that note should probably be only in the CLI section since users cannot skip creating/selecting mappings in the UI

Reviewer 2: BTW, choosing this solution was easier for us than totally skipping on the mapping steps in the wizard, but we can consider doing that with the refactored new dialogs if required.

Reviewer: > ... but we enhanced that as a feature to say that you can always create a plan with unmapped networks/storage, just to avoid blocking the plan creation, and it can be filled out later on once relevant.

to me this approach seems to go against the recent changes we've made in the UI to improve the user experience - as a user, I'd prefer the UI to block me from posting invalid input rather than posting it and then set a failure on the entity that would block me from going forward. in this case, I'd prefer to cancel the plan creation, create a mapping and then create the plan. btw, does it even work when the mapping is updated? AFAIR the UI clones the template mapping - do you go over the clones as well? @sgratch

Author: @sgratch @ahadas It sounds like there are two issues here. First, the phrase "empty mapping" isn't useful, so I can just remove the second sentence in the note. Second, do we want to call attention to this functionality? If not, it makes sense to remove the note altogether. But...maybe we want to leave the first sentence of the note in but add a warning such as "Be sure the mapping is still valid before attempting to implement it"? [Phrasing suggestions are welcome.]

Reviewer: @RichardHoch +1 on dropping the "empty mapping" part
about the other part - in general, our documentation should reflect how simple it is to migrate using MTV rather than describing all the features (and in this case, I'm not even sure if this feature makes sense as it raises few questions, e.g., what if I create a plan with "valid" mappings and then someone changes it and make the plan invalid - this is something we should avoid)

Reviewer 2: @ahadas @RichardHoch 

Regarding the current functionality discussion:
> to me this approach seems to go against the recent changes we've made in the UI to improve the user experience - as a user, I'd prefer the UI to block me from posting invalid input rather than posting it and then set a failure on the entity that would block me from going forward. in this case, I'd prefer to cancel the plan creation, create a mapping and then create the plan. 

Correct,  please remember that this plan wizard is a legacy code. For 2.5.0, the current wizard behavior avoids blocking plan creation for invalid/wrong/empty mappings, since the assumption was that we didn't want to block it. So the plan is created with an error status and a detailed tooltip of, e.g:

![Screenshot from 2023-09-11 16-17-51](https://github.com/kubev2v/forklift-documentation/assets/18169498/8e823759-c555-447a-955e-70a31d5d9621)

We'll need to handle that as part of the refactored new dialog/wizard. I'll suggest opening an issue on this.

> btw, does it even work when the mapping is updated? AFAIR the UI clones the template mapping - do you go over the cloned mappings as well? @sgratch

No we don't. It will be left as is.
In case you edit the plan, you might get a warning on the mapping step that the previous mapping is invalid, but it won't be related to the original template and it's not auto updated once the template is updated.

* *Source provider*: Select a source provider.
* *Target provider*: Select a target provider.

> . To create a storage mapping, click *Add* and map source disk storages to target storage classes as follows:

Reviewer 3: ```suggestion
. To create a storage mapping, click *Add* and map source disk storage to target storage classes as follows:
```

Unsure about this one… I think storage might convey singular or plural

Reviewer: yeah, github also doesn't like `storages`..
how about:
```suggestion
. To create a storage mapping, click *Add* and map disk storage sources to target storage classes as follows:
```

.. If your source provider is VMware vSphere, select a *Source datastore* and a *Target storage class*.
.. If your source provider is {rhv-full}, select a *Source storage domain* and a *Target storage class*.
.. If your source provider is {osp}, select a *Source volume type* and a *Target storage class*.
.. If your source provider is a set of one or more OVA files, select a *Source* and a *Target storage class* for the dummy storage that applies to all virtual disks within the OVA files.
.. If your storage provider is {virt}. select a *Source storage class* and a *Target storage class*.

> .. Optional: Click *Add* to create additional storage mappings or to map multiple source disk storages to a single target storage class.

Reviewer 3: ```suggestion
.. Optional: Click *Add* to create additional storage mappings or to map multiple source disk storage to a single target storage class.
```
*???* - unsure. WDYT
see above

. Click *Create StorageMap*.
. Specify the following fields:

> * *Name*: Enter a name to display in the storage mappings list. For example, `Default storage map` or `Empty storage mapping`.

Reviewer: also here I'd drop the "Empty storage mapping"

Author: Removed sentence as above.

* *Source provider*: Select a source provider.
* *Target provider*: Select a target provider.

. To create a storage mapping, do one of the following:

> ** To create an empty mapping, click *Create*.

Reviewer: and drop this part

.. If your source provider is VMware, select a *Source datastore* and a *Target storage class*.
.. If your source provider is {rhv-full}, select a *Source storage domain* and a *Target storage class*.
.. If your source provider is {osp}, select a *Source volume type* and a *Target storage class*.

> .. If your storage provider is {virt}. select a *Source provider* and a *Target provider*.

Reviewer: should be `Source storage class` and `Target storage class`, does it really refer to "provider" at the moment?

Author: ![ocp_ocp_storage map](https://github.com/kubev2v/forklift-documentation/assets/60698649/9a4f3405-7ad1-458a-9d36-7b4715ee5944)
This is what I saw using OCP providers @qiyuann created for me -- but maybe i did something wrong? Xin, what are you seeing for this step?

Reviewer 2: @RichardHoch two comments:

-  I suggest adding another step between step 3 to 4 - 

   After following step 3, the dialog looks as follows:

   ![Screenshot from 2023-08-16 21-23-17](https://github.com/kubev2v/forklift-documentation/assets/18169498/1674f7f9-8853-4b6e-8bee-58062012483d)

   There are 2 options for the user now:
    a. If he wants to create an **empty** mapping (i.e. declare the source and target providers but leave the specific storage mappings empty for future use) then jump to step 6 by clicking the **Create** button.

     b. If he wants to set the mapping(s) instead of leaving it empty, then in step 4 we need to add that the user should click the **Add** button to create the  first  storage mappings. 

   After clicking the **Add** button in (b), the dialog looks like this:

      ![Screenshot from 2023-08-16 21-31-50](https://github.com/kubev2v/forklift-documentation/assets/18169498/22ebfe88-a3cc-4d19-a823-3c95e8d682a3)

    **Please note** that this comment is not relevant specifically to OCP - OCP migration but to all providers and also seems relevant to the network mapping dialog (`step 4.3 Creating a network mapping`).

- Regarding step 4.d, it should be `Source storage class` and `Target storage class` as mentioned by Arik. 
  Currently there is a [bug](https://github.com/kubev2v/forklift-console-plugin/issues/691) since the `Source` label appears for the source instead of `Source storage class`. We'll fix it.

Reviewer 4: > This is what I saw using OCP providers @qiyuann created for me -- but maybe i did something wrong? Xin, what are you seeing for this step?

The Source/Target storage class could appear when the `Add` button is clicked.
